server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1917,7 +1919,40 @@ public interface Warmer {
      * @param translogRecoveryRunner the translog recovery runner
      * @param recoverUpToSeqNo       the upper bound, inclusive, of sequence number to be recovered
      */
-    public abstract Engine recoverFromTranslog(TranslogRecoveryRunner translogRecoveryRunner, long recoverUpToSeqNo) throws IOException;
+    // TODO move this blocking implementation into tests (adding a timeout) and make all the production usages async
+    public final void recoverFromTranslog(TranslogRecoveryRunner translogRecoveryRunner, long recoverUpToSeqNo) throws IOException {
+        final var future = new PlainActionFuture<Void>();
+        recoverFromTranslog(translogRecoveryRunner, recoverUpToSeqNo, future);
+        try {
+            future.get();
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof IOException ioException) {
```

Comment:
I don't have much intuition as to why all these catch'es?

Reply:
`Future#get` (and therefore `PlainActionFuture#get`) wraps exceptions thrown by the computation in an `ExecutionException`, to distinguish that kind of exception from an exception related to the `get()` call itself which may be interrupted. To maintain the same exception semantics as seen by callers, we must undo this wrapping. Since the original method is declared as `throws IOException` (and implicitly also `RuntimeException`) we must seprarately unwrap both `IOException` and `RuntimeException` to achieve this.

This is all temporary - future steps will move this blocking method up the stack until eventually it meets with another async method and evaporates, or else it vanishes into tests that do not care about this subtlety.

Reply:
I copied that ^ as comments in the code to help future readers - see 87b4c27aa78.